flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoder.java

```diff
@@ -0,0 +1,143 @@
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.MAGIC_NUMBER;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Decodes messages from the fragmentary netty buffers. This decoder assumes the
+ * messages have the following format:
+ * +-----------------------------------+--------------------------------+
+ * | FRAME_HEADER ||  MESSAGE_HEADER   |     DATA BUFFER (Optional)     |
+ * +-----------------------------------+--------------------------------+
+ *
+ * This decoder decodes the frame header and delegates the other following work
+ * to corresponding message parsers according to the message type. During the process
+ * of decoding, the decoder and parsers try best to eliminate copying. For the frame
+ * header and message header, it only cumulates data when they span multiple input buffers.
+ * For the buffer part, it copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class NettyMessageClientDecoder extends ChannelInboundHandlerAdapter {
+
+	/** The message parser for buffer response. */
+    private final NettyMessageDecoderDelegate bufferResponseDecoderDelegate;
+
+    /** The message parser for other messages other than buffer response. */
+	private final NettyMessageDecoderDelegate nonBufferResponseDecoderDelegate;
+
+	/** The cumulation buffer for the frame header part. */
+	private ByteBuf frameHeaderBuffer;
+
+	/**
+	 * The chosen message parser for the current message. If it is null, then
+	 * we are decoding the frame header part, otherwise we are decoding the actual
+	 * message.
+	 */
+	private NettyMessageDecoderDelegate currentDecoderDelegate;
+
+    NettyMessageClientDecoder(NetworkClientHandler networkClientHandler) {
+        this.bufferResponseDecoderDelegate = new BufferResponseDecoderDelegate(networkClientHandler);
+        this.nonBufferResponseDecoderDelegate = new NonBufferResponseDecoderDelegate();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        bufferResponseDecoderDelegate.onChannelActive(ctx.alloc());
+        nonBufferResponseDecoderDelegate.onChannelActive(ctx.alloc());
+
+		frameHeaderBuffer = ctx.alloc().directBuffer(FRAME_HEADER_LENGTH);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+		bufferResponseDecoderDelegate.release();
+		nonBufferResponseDecoderDelegate.release();
+
+		frameHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof ByteBuf)) {
+            ctx.fireChannelRead(msg);
```

Comment:
It must be `ByteBuf` in this stack, otherwise we should throw `Exception` instead of `fireChannelRead`?

Reply:
Agree with that and removed the `firChannelRead` for objects other than `ByteBuf`. Besides, I think we could omit the explicit checking here for performance? It will throw exception when casting the `Object` to `ByteBuf`.

Reply:
It would be friendly for judging before type transformation and i guess it has no performance concern here.

Reply:
After trying remove this part of code, I found that some tests are relying on the behavior that decoder would pass non-ByteBuf message directly. `NettyMessageDecoder` would also do so (inherit from the `ByteToMessageDecoder`). Thus we might first keep this behavior currently.